source/device.hpp

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include <set>
+#include <unordered_map>
+
+namespace sim {
+
+class Link;
+
+enum DeviceType { SWITCH, SENDER, RECEIVER };
+
+class Device {
+public:
+    Device(DeviceType a_type);
+    virtual ~Device();
+    virtual void process();
+    void add_inlink(Link *link);
+    void add_outlink(Device *dest, Link *link);
+
```

Comment:
Necessary methods for `Simulator::recalculate_paths()`

```suggestion
    void update_routing_table(Device *dest, Link *link);
    std::vector<Device *> get_neighbors() const;
    Link *get_link_to_device(Device *a_device) const;
```

Reply:
I agree about the routing table
```
void update_routing_table(Device *dest, Link *link);
```
but still uncertain about the rest. Does `get_neighbors()` return all neighbours or in/out neighbours? Do we really need `get_link_to_device`? Let's delegate these changes to https://github.com/Cloud-Storage-Team/algnet/pull/68.